app/views/questionnaire.py

```diff
@@ -586,15 +586,22 @@ def _remove_dependent_answers_from_completed_blocks(answer_id, questionnaire_sto
     :param questionnaire_store: holds the completed blocks
     :return: None
     """
+    answer_in_repeating_group = g.schema.answer_is_in_repeating_group(answer_id)
     dependencies = g.schema.dependencies[answer_id]
+
     for dependency in dependencies:
+        dependency_in_repeating_group = g.schema.answer_is_in_repeating_group(dependency)
+
         answer = g.schema.get_answer(dependency)
         question = g.schema.get_question(answer['parent_id'])
         block = g.schema.get_block(question['parent_id'])
 
-        location = Location(block['parent_id'], 0, block['id'])
-        if location in questionnaire_store.completed_blocks:
-            questionnaire_store.completed_blocks.remove(location)
+        if dependency_in_repeating_group and not answer_in_repeating_group:
+            questionnaire_store.remove_completed_blocks(group_id=block['parent_id'], block_id=block['id'])
+        else:
+            location = Location(block['parent_id'], group_instance, block['id'])
+            if location in questionnaire_store.completed_blocks:
```

Comment:
Not going to block on something like this... but would it not be better to have this conditional within the `remove_completed_blocks` method itself? It seems like a reasonable thing for the `QuestionnaireSchema` object to manage itself as it's only related to its internal state.

Reply:
Not sure . First of all I thought you were right , but its questionnaire.py which is a bunch of functions that have the business logic that use the schema and utalize the store .  I think we may be better off with a Questionnaire class that has a schema and a store ? I think its worth a talk through this